### PR TITLE
PICARD-2465: Fixed potential recursion error when loading many releases

### PR DESCRIPTION
**Pull request by phw**
# Summary

* This is a…
  * [x] Bug fix
  * [ ] Feature addition
  * [ ] Refactoring
  * [ ] Minor / simple change (like a typo)
  * [ ] Other
* **Describe this change in 1-2 sentences**:

# Problem

* JIRA ticket (_optional_): PICARD-2465

process_events_iter can cause unintended side effects. When loading many releases in parallel this can cause recursion depths errors.

This partially reverts changes done for PICARD-2454.

### CODE REVIEW

**Comment by phw** [comment]
This kind of emphasiszes that `processEvents` should be avoided if possible.

Just for referene some excerpt of the exceptions happening:

```
Original exception was:
Traceback (most recent call last):
  File "/home/phw/devel/musicbrainz/picard/./picard/tagger.py", line 436, in event
    event.run()
  File "/home/phw/devel/musicbrainz/picard/./picard/util/thread.py", line 50, in run
    self.func(*self.args, **self.kwargs)
  File "/home/phw/devel/musicbrainz/picard/./picard/file.py", line 280, in _loading_finished
    callback(self)
  File "/home/phw/devel/musicbrainz/picard/./picard/tagger.py", line 478, in _file_loaded
    self.move_file_to_track(file, albumid, recordingid)
  File "/home/phw/devel/musicbrainz/picard/./picard/tagger.py", line 379, in move_file_to_track
    album = self.load_album(albumid)
  File "/home/phw/devel/musicbrainz/picard/./picard/tagger.py", line 715, in load_album
    album = Album(album_id, discid=discid)
  File "/home/phw/devel/musicbrainz/picard/./picard/album.py", line 144, in __init__
    self.unmatched_files = Cluster(_("Unmatched Files"), special=True, related_album=self, hide_if_empty=True)
  File "/home/phw/devel/musicbrainz/picard/./picard/cluster.py", line 114, in __init__
    self.metadata['totaltracks'] = 0
  File "/home/phw/devel/musicbrainz/picard/./picard/metadata.py", line 471, in __setitem__
    self.set(name, values)
  File "/home/phw/devel/musicbrainz/picard/./picard/metadata.py", line 460, in set
    if isinstance(values, str) or not isinstance(values, Iterable):
  File "/usr/lib/python3.10/abc.py", line 119, in __instancecheck__
    return _abc_instancecheck(cls, instance)
RecursionError: maximum recursion depth exceeded in comparison
Traceback (most recent call last):
  File "/home/phw/devel/musicbrainz/picard/./picard/tagger.py", line 436, in event
    event.run()
  File "/home/phw/devel/musicbrainz/picard/./picard/util/thread.py", line 50, in run
    self.func(*self.args, **self.kwargs)
  File "/home/phw/devel/musicbrainz/picard/./picard/file.py", line 280, in _loading_finished
    callback(self)
  File "/home/phw/devel/musicbrainz/picard/./picard/tagger.py", line 478, in _file_loaded
    self.move_file_to_track(file, albumid, recordingid)
  File "/home/phw/devel/musicbrainz/picard/./picard/tagger.py", line 379, in move_file_to_track
    album = self.load_album(albumid)
  File "/home/phw/devel/musicbrainz/picard/./picard/tagger.py", line 715, in load_album
    album = Album(album_id, discid=discid)
  File "/home/phw/devel/musicbrainz/picard/./picard/album.py", line 144, in __init__
    self.unmatched_files = Cluster(_("Unmatched Files"), special=True, related_album=self, hide_if_empty=True)
  File "/home/phw/devel/musicbrainz/picard/./picard/cluster.py", line 114, in __init__
    self.metadata['totaltracks'] = 0
  File "/home/phw/devel/musicbrainz/picard/./picard/metadata.py", line 471, in __setitem__
    self.set(name, values)
  File "/home/phw/devel/musicbrainz/picard/./picard/metadata.py", line 460, in set
    if isinstance(values, str) or not isinstance(values, Iterable):
  File "/usr/lib/python3.10/abc.py", line 119, in __instancecheck__
    return _abc_instancecheck(cls, instance)
RecursionError: maximum recursion depth exceeded in comparison
Error in sys.excepthook:
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/apport_python_hook.py", line 72, in apport_excepthook
    from apport.fileutils import likely_packaged, get_recent_crashes
  File "/usr/lib/python3/dist-packages/apport/__init__.py", line 5, in <module>
    from apport.report import Report
  File "<frozen importlib._bootstrap>", line 1027, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1002, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 945, in _find_spec
  File "<frozen importlib._bootstrap_external>", line 1439, in find_spec
  File "<frozen importlib._bootstrap_external>", line 1406, in _get_spec
RecursionError: maximum recursion depth exceeded in __instancecheck__

Original exception was:
Traceback (most recent call last):
  File "/home/phw/devel/musicbrainz/picard/./picard/tagger.py", line 436, in event
    event.run()
  File "/home/phw/devel/musicbrainz/picard/./picard/util/thread.py", line 50, in run
    self.func(*self.args, **self.kwargs)
  File "/home/phw/devel/musicbrainz/picard/./picard/file.py", line 280, in _loading_finished
    callback(self)
  File "/home/phw/devel/musicbrainz/picard/./picard/tagger.py", line 478, in _file_loaded
    self.move_file_to_track(file, albumid, recordingid)
  File "/home/phw/devel/musicbrainz/picard/./picard/tagger.py", line 379, in move_file_to_track
    album = self.load_album(albumid)
  File "/home/phw/devel/musicbrainz/picard/./picard/tagger.py", line 715, in load_album
    album = Album(album_id, discid=discid)
  File "/home/phw/devel/musicbrainz/picard/./picard/album.py", line 144, in __init__
    self.unmatched_files = Cluster(_("Unmatched Files"), special=True, related_album=self, hide_if_empty=True)
  File "/home/phw/devel/musicbrainz/picard/./picard/cluster.py", line 114, in __init__
    self.metadata['totaltracks'] = 0
  File "/home/phw/devel/musicbrainz/picard/./picard/metadata.py", line 471, in __setitem__
    self.set(name, values)
  File "/home/phw/devel/musicbrainz/picard/./picard/metadata.py", line 460, in set
    if isinstance(values, str) or not isinstance(values, Iterable):
  File "/usr/lib/python3.10/abc.py", line 119, in __instancecheck__
    return _abc_instancecheck(cls, instance)
RecursionError: maximum recursion depth exceeded in comparison
Traceback (most recent call last):
  File "/home/phw/devel/musicbrainz/picard/./picard/tagger.py", line 436, in event
    event.run()
  File "/home/phw/devel/musicbrainz/picard/./picard/util/thread.py", line 50, in run
    self.func(*self.args, **self.kwargs)
  File "/home/phw/devel/musicbrainz/picard/./picard/file.py", line 280, in _loading_finished
    callback(self)
  File "/home/phw/devel/musicbrainz/picard/./picard/tagger.py", line 478, in _file_loaded
    self.move_file_to_track(file, albumid, recordingid)
  File "/home/phw/devel/musicbrainz/picard/./picard/tagger.py", line 379, in move_file_to_track
    album = self.load_album(albumid)
  File "/home/phw/devel/musicbrainz/picard/./picard/tagger.py", line 715, in load_album
    album = Album(album_id, discid=discid)
  File "/home/phw/devel/musicbrainz/picard/./picard/album.py", line 144, in __init__
    self.unmatched_files = Cluster(_("Unmatched Files"), special=True, related_album=self, hide_if_empty=True)
  File "/home/phw/devel/musicbrainz/picard/./picard/cluster.py", line 114, in __init__
    self.metadata['totaltracks'] = 0
  File "/home/phw/devel/musicbrainz/picard/./picard/metadata.py", line 471, in __setitem__
    self.set(name, values)
  File "/home/phw/devel/musicbrainz/picard/./picard/metadata.py", line 460, in set
    if isinstance(values, str) or not isinstance(values, Iterable):
  File "/usr/lib/python3.10/abc.py", line 119, in __instancecheck__
    return _abc_instancecheck(cls, instance)
RecursionError: maximum recursion depth exceeded in comparison
Error in sys.excepthook:
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/apport_python_hook.py", line 72, in apport_excepthook
    from apport.fileutils import likely_packaged, get_recent_crashes
  File "/usr/lib/python3/dist-packages/apport/__init__.py", line 5, in <module>
    from apport.report import Report
  File "<frozen importlib._bootstrap>", line 1027, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1002, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 945, in _find_spec
  File "<frozen importlib._bootstrap_external>", line 1439, in find_spec
  File "<frozen importlib._bootstrap_external>", line 1406, in _get_spec
RecursionError: maximum recursion depth exceeded in __instancecheck__

Original exception was:
Traceback (most recent call last):
  File "/home/phw/devel/musicbrainz/picard/./picard/tagger.py", line 436, in event
    event.run()
  File "/home/phw/devel/musicbrainz/picard/./picard/util/thread.py", line 50, in run
    self.func(*self.args, **self.kwargs)
  File "/home/phw/devel/musicbrainz/picard/./picard/file.py", line 280, in _loading_finished
    callback(self)
  File "/home/phw/devel/musicbrainz/picard/./picard/tagger.py", line 478, in _file_loaded
    self.move_file_to_track(file, albumid, recordingid)
  File "/home/phw/devel/musicbrainz/picard/./picard/tagger.py", line 379, in move_file_to_track
    album = self.load_album(albumid)
  File "/home/phw/devel/musicbrainz/picard/./picard/tagger.py", line 715, in load_album
    album = Album(album_id, discid=discid)
  File "/home/phw/devel/musicbrainz/picard/./picard/album.py", line 144, in __init__
    self.unmatched_files = Cluster(_("Unmatched Files"), special=True, related_album=self, hide_if_empty=True)
  File "/home/phw/devel/musicbrainz/picard/./picard/cluster.py", line 114, in __init__
    self.metadata['totaltracks'] = 0
  File "/home/phw/devel/musicbrainz/picard/./picard/metadata.py", line 471, in __setitem__
    self.set(name, values)
  File "/home/phw/devel/musicbrainz/picard/./picard/metadata.py", line 460, in set
    if isinstance(values, str) or not isinstance(values, Iterable):
  File "/usr/lib/python3.10/abc.py", line 119, in __instancecheck__
    return _abc_instancecheck(cls, instance)
RecursionError: maximum recursion depth exceeded in comparison
Traceback (most recent call last):
  File "/home/phw/devel/musicbrainz/picard/./picard/tagger.py", line 436, in event
    event.run()
  File "/home/phw/devel/musicbrainz/picard/./picard/util/thread.py", line 50, in run
    self.func(*self.args, **self.kwargs)
  File "/home/phw/devel/musicbrainz/picard/./picard/file.py", line 280, in _loading_finished
    callback(self)
  File "/home/phw/devel/musicbrainz/picard/./picard/tagger.py", line 478, in _file_loaded
    self.move_file_to_track(file, albumid, recordingid)
  File "/home/phw/devel/musicbrainz/picard/./picard/tagger.py", line 379, in move_file_to_track
    album = self.load_album(albumid)
  File "/home/phw/devel/musicbrainz/picard/./picard/tagger.py", line 715, in load_album
    album = Album(album_id, discid=discid)
  File "/home/phw/devel/musicbrainz/picard/./picard/album.py", line 144, in __init__
    self.unmatched_files = Cluster(_("Unmatched Files"), special=True, related_album=self, hide_if_empty=True)
  File "/home/phw/devel/musicbrainz/picard/./picard/cluster.py", line 114, in __init__
    self.metadata['totaltracks'] = 0
  File "/home/phw/devel/musicbrainz/picard/./picard/metadata.py", line 471, in __setitem__
    self.set(name, values)
  File "/home/phw/devel/musicbrainz/picard/./picard/metadata.py", line 460, in set
    if isinstance(values, str) or not isinstance(values, Iterable):
  File "/usr/lib/python3.10/abc.py", line 119, in __instancecheck__
    return _abc_instancecheck(cls, instance)
RecursionError: maximum recursion depth exceeded in comparison
Error in sys.excepthook:
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/apport_python_hook.py", line 72, in apport_excepthook
    from apport.fileutils import likely_packaged, get_recent_crashes
  File "/usr/lib/python3/dist-packages/apport/__init__.py", line 5, in <module>
    from apport.report import Report
  File "<frozen importlib._bootstrap>", line 1027, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1002, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 945, in _find_spec
  File "<frozen importlib._bootstrap_external>", line 1439, in find_spec
  File "<frozen importlib._bootstrap_external>", line 1406, in _get_spec
RecursionError: maximum recursion depth exceeded in __instancecheck__

Original exception was:
Traceback (most recent call last):
  File "/home/phw/devel/musicbrainz/picard/./picard/tagger.py", line 436, in event
    event.run()
  File "/home/phw/devel/musicbrainz/picard/./picard/util/thread.py", line 50, in run
    self.func(*self.args, **self.kwargs)
  File "/home/phw/devel/musicbrainz/picard/./picard/file.py", line 280, in _loading_finished
    callback(self)
  File "/home/phw/devel/musicbrainz/picard/./picard/tagger.py", line 478, in _file_loaded
    self.move_file_to_track(file, albumid, recordingid)
  File "/home/phw/devel/musicbrainz/picard/./picard/tagger.py", line 379, in move_file_to_track
    album = self.load_album(albumid)
  File "/home/phw/devel/musicbrainz/picard/./picard/tagger.py", line 715, in load_album
    album = Album(album_id, discid=discid)
  File "/home/phw/devel/musicbrainz/picard/./picard/album.py", line 144, in __init__
    self.unmatched_files = Cluster(_("Unmatched Files"), special=True, related_album=self, hide_if_empty=True)
  File "/home/phw/devel/musicbrainz/picard/./picard/cluster.py", line 114, in __init__
    self.metadata['totaltracks'] = 0
  File "/home/phw/devel/musicbrainz/picard/./picard/metadata.py", line 471, in __setitem__
    self.set(name, values)
  File "/home/phw/devel/musicbrainz/picard/./picard/metadata.py", line 460, in set
    if isinstance(values, str) or not isinstance(values, Iterable):
  File "/usr/lib/python3.10/abc.py", line 119, in __instancecheck__
    return _abc_instancecheck(cls, instance)
RecursionError: maximum recursion depth exceeded in comparison
Traceback (most recent call last):
  File "/home/phw/devel/musicbrainz/picard/./picard/tagger.py", line 436, in event
    event.run()
  File "/home/phw/devel/musicbrainz/picard/./picard/util/thread.py", line 50, in run
    self.func(*self.args, **self.kwargs)
  File "/home/phw/devel/musicbrainz/picard/./picard/file.py", line 280, in _loading_finished
    callback(self)
  File "/home/phw/devel/musicbrainz/picard/./picard/tagger.py", line 478, in _file_loaded
    self.move_file_to_track(file, albumid, recordingid)
  File "/home/phw/devel/musicbrainz/picard/./picard/tagger.py", line 379, in move_file_to_track
    album = self.load_album(albumid)
  File "/home/phw/devel/musicbrainz/picard/./picard/tagger.py", line 715, in load_album
    album = Album(album_id, discid=discid)
  File "/home/phw/devel/musicbrainz/picard/./picard/album.py", line 144, in __init__
    self.unmatched_files = Cluster(_("Unmatched Files"), special=True, related_album=self, hide_if_empty=True)
  File "/home/phw/devel/musicbrainz/picard/./picard/cluster.py", line 114, in __init__
    self.metadata['totaltracks'] = 0
  File "/home/phw/devel/musicbrainz/picard/./picard/metadata.py", line 471, in __setitem__
    self.set(name, values)
  File "/home/phw/devel/musicbrainz/picard/./picard/metadata.py", line 460, in set
    if isinstance(values, str) or not isinstance(values, Iterable):
  File "/usr/lib/python3.10/abc.py", line 119, in __instancecheck__
    return _abc_instancecheck(cls, instance)
RecursionError: maximum recursion depth exceeded in comparison
Error in sys.excepthook:
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/apport_python_hook.py", line 72, in apport_excepthook
    from apport.fileutils import likely_packaged, get_recent_crashes
  File "/usr/lib/python3/dist-packages/apport/__init__.py", line 5, in <module>
    from apport.report import Report
  File "<frozen importlib._bootstrap>", line 1027, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1002, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 945, in _find_spec
  File "<frozen importlib._bootstrap_external>", line 1439, in find_spec
  File "<frozen importlib._bootstrap_external>", line 1406, in _get_spec
RecursionError: maximum recursion depth exceeded in __instancecheck__

Original exception was:
Traceback (most recent call last):
  File "/home/phw/devel/musicbrainz/picard/./picard/tagger.py", line 436, in event
    event.run()
  File "/home/phw/devel/musicbrainz/picard/./picard/util/thread.py", line 50, in run
    self.func(*self.args, **self.kwargs)
  File "/home/phw/devel/musicbrainz/picard/./picard/file.py", line 280, in _loading_finished
    callback(self)
  File "/home/phw/devel/musicbrainz/picard/./picard/tagger.py", line 478, in _file_loaded
    self.move_file_to_track(file, albumid, recordingid)
  File "/home/phw/devel/musicbrainz/picard/./picard/tagger.py", line 379, in move_file_to_track
    album = self.load_album(albumid)
  File "/home/phw/devel/musicbrainz/picard/./picard/tagger.py", line 715, in load_album
    album = Album(album_id, discid=discid)
  File "/home/phw/devel/musicbrainz/picard/./picard/album.py", line 144, in __init__
    self.unmatched_files = Cluster(_("Unmatched Files"), special=True, related_album=self, hide_if_empty=True)
  File "/home/phw/devel/musicbrainz/picard/./picard/cluster.py", line 114, in __init__
    self.metadata['totaltracks'] = 0
  File "/home/phw/devel/musicbrainz/picard/./picard/metadata.py", line 471, in __setitem__
    self.set(name, values)
  File "/home/phw/devel/musicbrainz/picard/./picard/metadata.py", line 460, in set
    if isinstance(values, str) or not isinstance(values, Iterable):
  File "/usr/lib/python3.10/abc.py", line 119, in __instancecheck__
    return _abc_instancecheck(cls, instance)
RecursionError: maximum recursion depth exceeded in comparison
```

This happens only when loading many releases, but the resuls then are rather random.

The proper solution for this would be to run the moving in a thread, but that requires some serious refactoring. Currently moving a file between entities closely mixes the actual moving of the object and updating the involved object's data and triggering UI updates. What is needed is better separation of both, so the main work of updating the object's data (setting references, updating / merging metadata and images etc.) can be run in a thread with only the UI update being done on the main thread. The latter ideally in bulk for all related objects in one go.

**Comment by phw** [comment]
> Isn't there a risk that other calls to `process_events_iter()` lead to random crashs?

Using it somewhere could have side effects, but I don't think the remaining uses probably don't. We only use it at two places:

https://github.com/metabrainz/picard/blob/master/picard/tagger.py#L528-L550

`Tagger.move_files`: So far this seems to work ok. But this is exactly the use case of moving files multiple files again. We definitely should refactor this file moving.

https://github.com/metabrainz/picard/blob/master/picard/tagger.py#L945-L951

`Tagger._clustering_finished`: It's a similar case: Manye cluster creations + adding files to cluster in a loop. Maybe again we need to have this unified. The cluster creation and moving files should be done in one step inside the thread. Only when the objects have been created we trigger a single UI update on the main thread.

https://github.com/metabrainz/picard/blob/master/picard/album.py#L639-L647

`Album._match_files`: We have it still there in the code, but this is conditional and as far as I can tell we don't actually use it and could remove it.

Bottom line: For the 2.8 release reverting this process_events_iter should fix introduced issues and revert back to previous stability. The remaining uses of process_events_iter are luckily limited, but don't seem to cause (much) harm.

But yes, we definitely should start refactoring the handling of the `DataObject` subclasses and better separate the UI updates there. Especially the moving of files between those items must not mix up updating each of the three involved items (file, soure, target) and the UI update that much and must have the main part runnable in a thread.
